orientdb/src/main/java/org/hibernate/datastore/ogm/orientdb/OrientDBDialect.java

```diff
@@ -6,260 +6,165 @@
-		log.info( "overrideType:" + type.getName() + ";" + type.getReturnedClass() );
-		GridType gridType = null;
-		if ( type.getName().equals( "com.orientechnologies.orient.core.id.ORecordId" ) ) {
-			gridType = ORecordIdGridType.INSTANCE;
-		}
-		else {
-			gridType = super.overrideType( type ); // To change body of generated methods, choose Tools | Templates.
-		}
-		return gridType;
-	}
+public class OrientDBDialect extends BaseGridDialect {
+
+    private static final Log log = LoggerFactory.getLogger();
+    private OrientDBDatastoreProvider provider;
+
+    public OrientDBDialect(OrientDBDatastoreProvider provider) {
+        this.provider = provider;
+    }
+
+
+    public ODocument findRecordByKey(ODatabaseDocumentTx executionEngine, EntityKey key, TupleContext tupleContext) {
+        //TODO: include tupleContext selectableColumns projection ??
+        OSQLSynchQuery<ODocument> orientQuery = OrientDBQueryHelper.createSelect(key);
+        return executionEngine.command(orientQuery).execute();
+    }
+
+    private ODocument prepareObjectWithPk(EntityKey key) {
+        ODocument object = new ODocument();
+        String[] columnNames = key.getColumnNames();
+        Object[] columnValues = key.getColumnValues();
+        for (int i = 0; i < columnNames.length; i++) {
+            object.field(columnNames[i], columnValues[i]);
+        }
+        return object;
+    }
+
+    @Override
+    public Tuple getTuple(EntityKey key, TupleContext tupleContext) {
+        ODocument document = findRecordByKey(provider.getConnection(), key, tupleContext);
+        OrientDBTupleSnapshot tuple;
+        if (document == null) {
+            tuple = new OrientDBTupleSnapshot(prepareObjectWithPk(key), key.getMetadata(), OrientDBTupleSnapshot.SnapshotType.INSERT);
+        } else if (isInTheQueue(key, tupleContext)) {
+            tuple = new OrientDBTupleSnapshot(document, key.getMetadata(), OrientDBTupleSnapshot.SnapshotType.UPDATE);
+        } else {
+            return null;
+        }
+        return new Tuple(tuple);
+    }
+
+    private boolean isInTheQueue(EntityKey key, TupleContext tupleContext) {
+        OperationsQueue queue = tupleContext.getOperationsQueue();
+        return queue != null && queue.contains(key);
+    }
+
+    @Override
+    public Tuple createTuple(EntityKey key, TupleContext tupleContext) {
+        log.info("createTuple:EntityKey:" + key + "; tupleContext" + tupleContext);
+        ODocument newObject = prepareObjectWithPk(key);
+        return new Tuple(new OrientDBTupleSnapshot(newObject, key.getMetadata(), OrientDBTupleSnapshot.SnapshotType.INSERT));
+    }
+
+
+    @Override
+    public void insertOrUpdateTuple(EntityKey key, Tuple tuple, TupleContext tupleContext) throws TupleAlreadyExistsException {
+        log.info("insertOrUpdateTuple:EntityKey:" + key + "; tupleContext" + tupleContext + "; tuple:" + tuple);
+        ODatabaseDocumentTx connection = provider.getConnection();
```

Comment:
It is transactional, it's the default connection type for OrientDB, but it could be not transactional as well, which one should we use here ?